databuilder/rest_api/rest_api_query.py

```diff
@@ -0,0 +1,198 @@
+             URL after compute: http://foo.bar/dashboard/foo
+
+        :param params: A keyword arguments that pass into requests.get function.
+            https://requests.readthedocs.io/en/master/user/quickstart/#make-a-request
+        :param json_path: A JSONPath expression. https://github.com/kennknowles/python-jsonpath-rw
+        Example:
+            JSON result:
+            [{"report_id": "1", "report_name": "first report", "foo": "bar"},
+             {"report_id": "2", "report_name": "second report"}]
+
+            JSON PATH:
+            [*].[report_id,report_name]
+
+            ["1", "first report", "2", "second report"]
+        :param field_names: Field names to be used on the result. Result is dictionary where field_name will be the key
+         and the values extracted via JSON PATH will be the value.
+
+            JSON Path result:
+                ["1", "first report", "2", "second report"]
+
+            field_names:
+                ["dashboard_id", "dashboard_description"]
+
+            {"dashboard_id": "1", "dashboard_description": "first report"}
+            {"dashboard_id": "2", "dashboard_description": "second report"}
+
+        :param fail_no_result: If there's no result from the query it will make it fail.
+        :param skip_no_result: If there's no result from the query, it will skip this record.
+        """
+        self._inner_rest_api_query = query_to_join
+        self._url = url
+        self._params = params
+        self._json_path = json_path
+        self._jsonpath_expr = parse(self._json_path)
+        self._fail_no_result = fail_no_result
+        self._skip_no_result = skip_no_result
+        self._field_names = field_names
+        self._more_pages = False
+
+    def execute(self):
+        self._authenticate()
+
+        for record_dict in self._inner_rest_api_query.execute():
+
+            first_try = True  # To control pagination. Always pass the while loop on the first try
+            while first_try or self._more_pages:
+                first_try = False
+
+                url = self._preprocess_url(record=record_dict)
+                response = self._send_request(url=url)
+
+                response_json = response.json()  # type: Union[List[Any], Dict[str, Any]]
+
+                # value extraction via JSON Path
+                result_list = [match.value for match in self._jsonpath_expr.find(response_json)]  # type: List[Any]
+
+                if not result_list:
+                    log_msg = 'No result from URL: {url}  , JSONPATH: {json_path} , response payload: {response}' \
+                        .format(url=self._url, json_path=self._json_path, response=response_json)
+                    LOGGER.info(log_msg)
+
+                    if self._fail_no_result:
+                        raise Exception(log_msg)
+
+                    if self._skip_no_result:
+                        continue
+
+                    yield copy.deepcopy(record_dict)
```

Comment:
if there is not match result based on the required field, it just return the original result, is that what this line does?

Reply:
Yes, if `original` record means the one from previous RestAPIQuery. 

```
Previous record: {"foo": "bar}
result_list: []

--> yield {"foo": "bar}
```